p2p/discovery.py

```diff
@@ -234,9 +247,134 @@ def send_neighbours(self, node: kademlia.Node, neighbours: List[kademlia.Node])
             self.send(node, message)
 
 
+class PreferredNodeDiscoveryProtocol(DiscoveryProtocol):
+    """
+    A DiscoveryProtocol which has a list of preferred nodes which it will prioritize using before
+    trying to find nodes.  Each preferred node can only be used once every
+    preferred_node_recycle_time seconds.
+    """
+    preferred_nodes: Sequence[kademlia.Node] = None
+    preferred_node_recycle_time: int = 300
+    _preferred_node_tracker: Dict[kademlia.Node, float] = None
+
+    def __init__(self,
+                 privkey: datatypes.PrivateKey,
+                 address: kademlia.Address,
+                 bootstrap_nodes: Tuple[kademlia.Node, ...],
+                 preferred_nodes: Sequence[kademlia.Node] = None) -> None:
+        super().__init__(privkey, address, bootstrap_nodes)
+
+        if preferred_nodes is not None:
+            self.preferred_nodes = preferred_nodes
+        else:
+            self.logger.debug('Operating with no preferred nodes')
```

Comment:
Might be worth going ahead and doing a debug log in the other conditional branch to log what preferred nodes we are using.

Reply:
I'll move it outside of the if block, always logging self.preferred_nodes, and change it into a INFO as this is only going to show up once at startup